src/gfx/convert.cpp

```diff
@@ -0,0 +1,831 @@
+	 */
+	class RawTile {
+		std::array<std::array<size_t, 8>, 8> _pixelIndices{};
+
+	public:
+		// Not super clean, but it's closer to matrix notation
+		size_t &operator()(size_t x, size_t y) { return _pixelIndices[y][x]; }
+	};
+
+private:
+	std::vector<RawTile> _tiles;
+
+public:
+	/**
+	 * Creates a new raw tile, and returns a reference to it so it can be filled in
+	 */
+	RawTile &newTile() {
+		_tiles.emplace_back();
+		return _tiles.back();
+	}
+};
+
+struct AttrmapEntry {
+	size_t protoPaletteID;
+	uint16_t tileID;
+	bool yFlip;
+	bool xFlip;
+};
+
+static void outputPalettes(std::vector<Palette> const &palettes) {
+	std::filebuf output;
+	output.open(options.palettes, std::ios_base::out | std::ios_base::binary);
+
+	for (Palette const &palette : palettes) {
+		for (uint16_t color : palette) {
+			output.sputc(color & 0xFF);
+			output.sputc(color >> 8);
+		}
+	}
+}
+
+namespace unoptimized {
+
+// TODO: this is very redundant with `TileData`; try merging both?
+static void outputTileData(Png const &png, DefaultInitVec<AttrmapEntry> const &attrmap,
+                           std::vector<Palette> const &palettes,
+                           DefaultInitVec<size_t> const &mappings) {
+	std::filebuf output;
+	output.open(options.output, std::ios_base::out | std::ios_base::binary);
+
+	auto iter = attrmap.begin();
+	for (auto tile : png.visitAsTiles(options.columnMajor)) {
+		Palette const &palette = palettes[mappings[iter->protoPaletteID]];
+		for (uint32_t y = 0; y < 8; ++y) {
+			uint16_t row = 0;
+			for (uint32_t x = 0; x < 8; ++x) {
+				row <<= 1;
+				uint8_t index = palette.indexOf(tile.pixel(x, y));
+				if (index & 1) {
+					row |= 0x001;
+				}
+				if (index & 2) {
+					row |= 0x100;
+				}
+			}
+			output.sputc(row & 0xFF);
+			output.sputc(row >> 8);
+		}
+		++iter;
+	}
+	assert(iter == attrmap.end());
+}
+
+static void outputMaps(Png const &png, DefaultInitVec<AttrmapEntry> const &attrmap,
+                       DefaultInitVec<size_t> const &mappings) {
+	std::optional<std::filebuf> tilemapOutput, attrmapOutput;
+	if (!options.tilemap.empty()) {
+		tilemapOutput.emplace();
+		tilemapOutput->open(options.tilemap, std::ios_base::out | std::ios_base::binary);
+	}
+	if (!options.attrmap.empty()) {
+		attrmapOutput.emplace();
+		attrmapOutput->open(options.attrmap, std::ios_base::out | std::ios_base::binary);
+	}
+
+	uint8_t tileID = 0;
+	uint8_t bank = 0;
+	auto iter = attrmap.begin();
+	for ([[maybe_unused]] auto tile : png.visitAsTiles(options.columnMajor)) {
+		if (tileID == options.maxNbTiles[bank]) {
+			assert(bank == 0);
+			bank = 1;
+			tileID = 0;
+		}
+
+		if (tilemapOutput.has_value()) {
+			tilemapOutput->sputc(tileID + options.baseTileIDs[bank]);
+		}
+		if (attrmapOutput.has_value()) {
+			uint8_t palID = mappings[iter->protoPaletteID] & 7;
+			attrmapOutput->sputc(palID | bank << 3); // The other flags are all 0
+			++iter;
+		}
+		++tileID;
+	}
+}
```

Comment:
Would this be correct?
```suggestion
	}
	assert(iter == attrmap.end());
}
```